### ISSUE TITLE
std::bad_alloc on executing query

### ISSUE THREAD

**Issue by epluribusunix** [arangodb/arangodb#7930]
## My Environment

* __ArangoDB Version__:        3.4.1
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Cluster
* __Deployment Strategy__:   Manual Start in Docker
* __Configuration__:               4x hosts (2x DB, 2x Coord)
* __Infrastructure__:               Private, VMware cluster
* __Operating System__:        CentOS 7.5
* __Total RAM in your machine__:        16 GB
* __Disks in use__:        Flash/SSD SAN
* __Used Package__:              Docker - official Docker library

## Component, Query & Data

__Affected feature__:
AQL query using web interface or with ArangoJS driver

__AQL query (if applicable)__:

```
LET matches = (
    FOR c IN BlueCat
        FOR p in BlueCat
            FILTER p._key == c.parent_id
            RETURN CONCAT_SEPARATOR('-', p._key, c._key)
)
LET edges = (
    FOR e IN BlueCat_edges
        RETURN e._key
)
LET inserts = MINUS(matches, edges)
FOR c IN BlueCat
    LET sub = LEFT(c.objType, 3)
    LET objType = (sub == 'IP4') ? 'IP4 Space' :
                  (sub == 'Dns') ? 'DNS' :
                  (sub == 'Tag') ? 'Tags' :
                  (sub == 'CON') ? 'CONF' : c.objType
    FOR p in BlueCat
        FILTER p._key == c.parent_id
        LET key = CONCAT_SEPARATOR('-', p._key, c._key)
        FOR i IN inserts
            FILTER i == key
            RETURN { _key: key, _from: p._id, _to: c._id, type: objType }
```

__AQL explain (if applicable)__:

```
Execution plan:
 Id   NodeType                  Site      Est.   Comment
  1   SingletonNode             COOR         1   * ROOT
  9   SubqueryNode              COOR         1     - LET matches = ...   /* const subquery */
  2   SingletonNode             DBS          1       * ROOT
  3   EnumerateCollectionNode   DBS     297985         - FOR c IN BlueCat   /* full collection scan, 8 shard(s) */
 44   RemoteNode                COOR    297985           - REMOTE
 45   GatherNode                COOR    297985           - GATHER 
 38   ScatterNode               COOR    297985           - SCATTER
 39   RemoteNode                DBS     297985           - REMOTE
 29   IndexNode                 DBS     297985           - FOR p IN BlueCat   /* primary index scan, index only, projections: `_key`, 8 shard(s) */
  7   CalculationNode           DBS     297985             - LET #18 = CONCAT_SEPARATOR("-", p.`_key`, c.`_key`)   /* simple expression */   /* collections used: p : BlueCat, c : BlueCat */
 40   RemoteNode                COOR    297985             - REMOTE
 41   GatherNode                COOR    297985             - GATHER 
  8   ReturnNode                COOR    297985             - RETURN #18
 14   SubqueryNode              COOR         1     - LET edges = ...   /* const subquery */
 10   SingletonNode             DBS          1       * ROOT
 11   EnumerateCollectionNode   DBS        801         - FOR e IN BlueCat_edges   /* full collection scan, projections: `_key`, 8 shard(s) */
 12   CalculationNode           DBS        801           - LET #20 = e.`_key`   /* attribute expression */   /* collections used: e : BlueCat_edges */
 48   RemoteNode                COOR       801           - REMOTE
 49   GatherNode                COOR       801           - GATHER 
 13   ReturnNode                COOR       801           - RETURN #20
 15   CalculationNode           COOR         1     - LET inserts = MINUS(matches, edges)   /* simple expression */
 34   ScatterNode               COOR         1     - SCATTER
 35   RemoteNode                DBS          1     - REMOTE
 16   EnumerateCollectionNode   DBS     297985     - FOR c IN BlueCat   /* full collection scan, 8 shard(s) */
 17   CalculationNode           DBS     297985       - LET sub = LEFT(c.`objType`, 3)   /* simple expression */   /* collections used: c : BlueCat */
 36   RemoteNode                COOR    297985       - REMOTE
 37   GatherNode                COOR    297985       - GATHER 
 30   ScatterNode               COOR    297985       - SCATTER
 31   RemoteNode                DBS     297985       - REMOTE
 28   IndexNode                 DBS     297985       - FOR p IN BlueCat   /* primary index scan, projections: `_id`, `_key`, 8 shard(s) */
 22   CalculationNode           DBS     297985         - LET key = CONCAT_SEPARATOR("-", p.`_key`, c.`_key`)   /* simple expression */   /* collections used: p : BlueCat, c : BlueCat */
 26   CalculationNode           DBS     297985         - LET #26 = { "_key" : key, "_from" : p.`_id`, "_to" : c.`_id`, "type" : ((sub == "IP4") ? "IP4 Space" : ((sub == "Dns") ? "DNS" : ((sub == "Tag") ? "Tags" : ((sub == "CON") ? "CONF" : c.`objType`)))) }   /* simple expression */   /* collections used: p : BlueCat, c : BlueCat */
 32   RemoteNode                COOR    297985         - REMOTE
 33   GatherNode                COOR    297985         - GATHER 
 23   EnumerateListNode         COOR  29798500         - FOR i IN inserts   /* list iteration */
 24   CalculationNode           COOR  29798500           - LET #24 = (i == key)   /* simple expression */
 25   FilterNode                COOR  29798500           - FILTER #24
 27   ReturnNode                COOR  29798500           - RETURN #26

Indexes used:
 By   Type      Collection   Unique   Sparse   Selectivity   Fields       Ranges
 29   primary   BlueCat      true     false       100.00 %   [ `_key` ]   (p.`_key` == c.`parent_id`)
 28   primary   BlueCat      true     false       100.00 %   [ `_key` ]   (p.`_key` == c.`parent_id`)

Functions used:
 Name               Deterministic   Cacheable   Uses V8
 CONCAT_SEPARATOR   true            true        false  
 MINUS              true            true        false  
 LEFT               true            true        false  

Optimization rules applied:
 Id   RuleName
  1   move-calculations-up
  2   move-filters-up
  3   remove-unnecessary-calculations
  4   move-calculations-up-2
  5   move-filters-up-2
  6   use-indexes
  7   remove-filter-covered-by-index
  8   remove-unnecessary-calculations-2
  9   scatter-in-cluster
 10   distribute-filtercalc-to-cluster
 11   remove-unnecessary-remote-scatter
 12   reduce-extraction-to-projection
```

__Dataset__:
The data describes our IP network and DNS structure, so it follows a familiar CIDR/DNS tree format.  There are about 300K documents in the collection, each quite small (normally 3-6 non-system attributes each, some as high as 12).

I cannot share dataset (_yet... I'm working on this_), but here is the structure:

```
{
  "allInSync": true,
  "indexes": [
    {
      "id": "31399110",
      "type": "hash",
      "fields": [
        "parent_id"
      ],
      "unique": false,
      "sparse": false,
      "deduplicate": true
    },
    {
      "id": "31399111",
      "type": "hash",
      "fields": [
        "objType"
      ],
      "unique": false,
      "sparse": false,
      "deduplicate": true
    },
    {
      "id": "31399112",
      "type": "hash",
      "fields": [
        "name"
      ],
      "unique": false,
      "sparse": false,
      "deduplicate": true
    },
    {
      "id": "31399113",
      "type": "hash",
      "fields": [
        "abs_name"
      ],
      "unique": false,
      "sparse": false,
      "deduplicate": true
    },
    {
      "id": "31399114",
      "type": "hash",
      "fields": [
        "ip_dd"
      ],
      "unique": false,
      "sparse": false,
      "deduplicate": true
    },
    {
      "id": "31399115",
      "type": "hash",
      "fields": [
        "objType",
        "ip_dd"
      ],
      "unique": false,
      "sparse": false,
      "deduplicate": true
    }
  ],
  "isReady": true,
  "parameters": {
    "cacheEnabled": false,
    "deleted": false,
    "globallyUniqueId": "c6013395/",
    "id": "6013395",
    "isSmart": false,
    "isSystem": false,
    "keyOptions": {
      "allowUserKeys": true,
      "type": "traditional"
    },
    "name": "BlueCat",
    "numberOfShards": 8,
    "planId": "6013395",
    "replicationFactor": 2,
    "shardKeys": [
      "_key"
    ],
    "shardingStrategy": "hash",
    "shards": {
      "s6013402": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ],
      "s6013401": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ],
      "s6013400": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ],
      "s6013403": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ],
      "s6013396": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ],
      "s6013397": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ],
      "s6013398": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ],
      "s6013399": [
        "PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4",
        "PRMR-31581e0f-198a-4d73-ba0f-b9142d305ccf"
      ]
    },
    "status": 3,
    "type": 2,
    "waitForSync": false
  },
  "planVersion": 784
}
```

__Size of your Dataset on disk__:
50 MB

__Replication Factor & Number of Shards (Cluster only)__:
"replicationFactor": 2,
"numberOfShards": 8,

## Steps to reproduce
I am trying to create edges for an imported dataset, but get an error running the query.  Originally, I thought it was related to import/upsert commands I was using, but it happens when only running a query.

This problem arose when I was trying to tune a query that worked well on a single server, but timed-out on the cluster.  Over several iterations, this one performed the best, but that doesn't mean it's the _right_ way to do it.  Coincidentally, I did not get the `std:bad_alloc` error during tuning/testing, probably because I limited the initial set to 10000 records.

I started with 8GB of memory on each server, and doubled that during testing.  Also, I have adjusted memory management configuration as such:

- /proc/sys/vm/overcommit_memory:  2
- /proc/sys/vm/overcommit_ratio:  90  (tested w/ 50 as well)
- /sys/kernel/mm/transparent_hugepage/enabled:  never
- /sys/kernel/mm/transparent_hugepage/defrag:  never

__Problem__:
The only error message appears in the browser window (no other errors appear in logs, that I can find):

> Query: AQL: Error message received from shard '' on cluster node 'PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4': Error message received from shard '' on cluster node 'CRDN-1b814b2a-508b-488d-aea3-c0069cbfe176': Error message received from shard '' on cluster node 'PRMR-e5962f7e-00a0-44fc-8106-136f466e58d4': Error message received from shard '' on cluster node 'CRDN-1b814b2a-508b-488d-aea3-c0069cbfe176': std::bad_alloc (exception location: /work/ArangoDB/arangod/Aql/ClusterBlocks.cpp:1018). Please report this error to arangodb.com (exception location: /work/ArangoDB/arangod/Aql/ClusterBlocks.cpp:1018). Please report this error to arangodb.com (exception location: /work/ArangoDB/arangod/Aql/ClusterBlocks.cpp:1018). Please report this error to arangodb.com (exception location: /work/ArangoDB/arangod/Aql/ClusterBlocks.cpp:1018). Please report this error to arangodb.com (while executing) (exception location: /work/ArangoDB/arangod/RestHandler/RestCursorHandler.cpp:238). Please report this error to arangodb.com

__Expected result__:
An array of edge documents like { _key, _from, _to, type }

**Comment by epluribusunix** [comment]
Additionally, I am getting the `std:bad_alloc` error when doing other operations as well (insert, delete, and replace), so the problem is not limited specifically to this query.

**Comment by kunalak** [comment]
Hi @epluribusunix Can you check if setting the overcommit_memory setting to 0 helps?(https://docs.arangodb.com/devel/Manual/Installation/LinuxOSTuningScripts.html)

**Comment by epluribusunix** [comment]
@kunalak, I tried setting overcommit_memory = `1`, with no luck, but am having more success with `0`.  So far, this is the query that has worked, although it took several minutes to complete:

`LET edges = ( FOR e IN BlueCat_edges RETURN e._key )`
`FOR c IN BlueCat`
`	FOR p in BlueCat`
`		FILTER p._key == c.parent_id`
`		LET key = CONCAT_SEPARATOR('-', p._key, c._key)`
`		FILTER POSITION(edges, key) == false`
`		LET sub = LEFT(c.objType, 3)`
`		LET objType = (sub == 'IP4') ? 'IP4 Space' :`
`                      (sub == 'Dns') ? 'DNS' :`
`                      (sub == 'Tag') ? 'Tags' :`
`                      (sub == 'CON') ? 'CONF' : c.objType`
`        INSERT { _key: key, _from: p._id, _to: c._id, type: objType }`
`        INTO BlueCat_edges`

I am running more tests, trying different ways to optimize the code, and will get back to you as soon as I can.  Also, I was able to anonymize the data, so I will try uploading a dataset shortly.

**Comment by epluribusunix** [comment]
@dothebart, thank you for your reply.  I am still coming to terms with how the cluster works - how queries are distributed and how to make the optimizer work **_for_** me.  I definitely understand the need for indexes, and that is specifically why the query is written this way.

From what I understand, the `_key` attribute is always indexed, and duplicating this would be counter-productive.  I then added an index for `parent_id` so that both attributes of the join are covered, and even tried adding an index with both keys (with no luck).  Even still, this query can be simplified down to it's most basic components, and still takes a LONG time.  This does no calculation, matches on two indexed fields, but takes 12+ seconds to return:
```
FOR c IN BlueCat
	FOR p in BlueCat
	    FILTER p._key == c.parent_id
	    RETURN { p: p._key, c: c._key }
```
```
Execution plan:
 Id   NodeType                  Site  Calls     Items   Runtime [s]   Comment
  1   SingletonNode             DBS       8         8       0.00004   * ROOT
  2   EnumerateCollectionNode   DBS     318    297581       0.11913     - FOR c IN BlueCat   /* full collection scan, 8 shard(s) */
 15   RemoteNode                COOR    636    297581       0.82958       - REMOTE
 16   GatherNode                COOR    629    297581       0.00283       - GATHER 
  9   ScatterNode               COOR   2757   2380648       0.87829       - SCATTER
 10   RemoteNode                DBS    4878   2380648       5.65669       - REMOTE
  8   IndexNode                 DBS    2665   2641183       3.82221       - FOR p IN BlueCat   /* primary index scan, index only, projections: '_key', 8 shard(s) */
  6   CalculationNode           DBS    2673    218693       0.17454         - LET #4 = { "p" : p.'_key', "c" : c.'_key' }   /* simple expression */   /* collections used: p : BlueCat, c : BlueCat */
 11   RemoteNode                COOR    468    218693       0.42148         - REMOTE
 12   GatherNode                COOR    461    218693       0.00208         - GATHER
  7   ReturnNode                COOR    453    218693       0.00316         - RETURN #4

Indexes used:
 By   Type      Collection   Unique   Sparse   Selectivity   Fields       Ranges
  8   primary   BlueCat      true     false       100.00 %   [ '_key' ]   (p.'_key' == c.'parent_id')

Query Profile:`
 Query Stage           Duration [s]`
 initializing               0.00001`
 parsing                    0.00006`
 optimizing ast             0.00001`
 loading collections        0.00000`
 instantiating plan         0.00002`
 optimizing plan            0.01095`
 executing                 12.25356`
 finalizing                 0.06814`
```

What I don't understand is why the query optimizer is not using the `parent_id` index:

Is this because it does not need an index for `parent_id` because this is a field of the current `FOR` document (therefore does not need to be looked-up)?

On a single server, this runs in a few milliseconds.  Since I can't easily change the data set (it's coming from an external source), how can this query be optimized?  Should I use a Foxx service to do all the joins outside of the DB engine?  Do all edge calculations need to be handled offline?

This question landed on GitHub because I was getting allocation errors using the recommended memory settings while doing (what I believed to be) simple joins.  If this is just not how the database is intended to be used, I need to know.  Is there a problem with my cluster setup (too-many/not-enough shards, etc.) or is this problem seen on all clusters?

**Comment by kunalak** [comment]
@epluribusunix How long did the query take to finish (from both the JS driver and the webUI)? (and how much much was the Query throughput)?  Also- Can you share a short summary of the query itself describing what the query aims to achieve? (to see if the data model/schemas could be designed more optimally with the right possible indexes)

**Comment by dothebart** [comment]
Hi, 
the outer `for` still is a `EnumerateCollectionNode` - which means full collection scan - instead of an `IndexNode`. The reason is, that you don't have any `FILTER` condition for `c`. However, you do have it for `p` - but the inner loop is executed for each `c`. This means your result is going to be `Nc` x `Np` - so your result set basically squares the amount of documents you have in `BlueCat` - which also explains why it runs out of memory.

Please have a look at: 
https://docs.arangodb.com/3.4/AQL/Examples/Join.html

**Comment by epluribusunix** [comment]
@kunalak - the query is creating edges based on the parent/child IDs of each document.  Documents are removed using the graph API, but when new vertices are added, I need to be able to create edges to existing parents (and children).  I can do this several ways, but each of them requires a join or lookup in an existing index and many of them require storing keys in arrays (and comparing using MINUS or POSITION).

Also, I have been running these queries from the Web UI, since its easier to iterate/debug than the code-based method.  I don't have throughput numbers, other than what is given when running `profile`.

@dothebart - thank you for the detail.  I realize this seems an N-squared issue, but the `c.parent_id = p._key` should be almost 1:1, so we're really only scanning the `c` collection once, and the `p` index (`_key`) Nc-times.  But since I'm creating edges between virtually all vertices, I really **_do_** want to scan all elements.

**I think the take-away here is that the database was not designed to do this sort of thing and I need to find a completely different approach.**  I will move edge calculation to offline JavaScript and re-write the query to do edge creations on a targeted set of vertices.

Thank you both for your guidance.

**Comment by dothebart** [comment]
Hi, 
in order to get it working you need to do it in chunks. Add a `SORT` to the outer `FOR` of the joins on an indexed property in combination with a `LIMIT` statement.  Use a chunk-size that is apropriate for your environment, like a 1000 on each go.

**Comment by kunalak** [comment]
@epluribusunix Is this query more like a maintenance operation? (done say like X times in a day or once every week?) ...In other words, what are the normal ongoing queries to the database? and which ones would be the infrequent/maintenance ones? (since both normally have different performance/latency expectations)

**Comment by dothebart** [comment]
So I get that correctly that you want the database to "refresh" in 15 minute intervals? 

Shouldn't the [outer join as sugested in the documentation](https://docs.arangodb.com/3.4/AQL/Examples/Join.html#outer-joins) do the same as your minus operation? Maybe writing the `_key`s of these into another temporary collection is a good idea, so you can work on them in chunks later on. 

Another question regarding your dataset size. From the figures you named it seems that this should fit on a single machine? On a single machine joins are much cheaper. If you want the cluster for resillience, https://docs.arangodb.com/3.4/Manual/Architecture/DeploymentModes/MasterSlave/ may be the better option?

**Comment by epluribusunix** [comment]
@dothebart, That outer join method was originally how I had written my queries, but found that performance suffers on a cluster due to the same N-squared issue, but with HTTP in the middle ([here is a link](https://stackoverflow.com/questions/53966083/arangodb-slow-query-performance-on-cluster) to the post on SO).

As for the necessity of a cluster, you might be right.  My datasets are not giant (8-10 collections of a few hundred thousand documents each), but I expect a lot of lookup traffic and need the system to be as reliable as possible.  I chose cluster mode so that I could employ multiple Foxx servers while enjoying the benefit of constantly-replicated database services.

When I talked with @kunalak last year, he also suggested a Master/Slave setup, but I'm wondering if an [Active Failover](https://docs.arangodb.com/3.4/Manual/Architecture/DeploymentModes/ActiveFailover/Architecture.html) would be better.  I **_really_** _do not_ want to implement custom replication code/processes, which seems to be the major detractor to the Master/Slave setup, but I could be reading the docs incorrectly.

Also, does sharding affect memory usage within the database engine?  I know how it works from a MongoDB perspective, but I can't find any information regarding benefits/detractors or suggested shard configurations beyond a few mentions to `numberOfShards` ([here](https://docs.arangodb.com/3.4/Manual/DataModeling/Collections/CollectionMethods.html) and [here](https://docs.arangodb.com/3.4/Manual/DataModeling/Collections/DatabaseMethods.html)).  Is there a benefit to using 4 shards (or _one_) instead of 8 or 16?

**Comment by kunalak** [comment]
@epluribusunix Active Failover is the right setup if you do need "Resilience" but your total dataset size fits on 1 machine. In other words, with Active Failover setup- your operations and data continue to be served from your primary server but the follower machine will asynchronously continue to replicate the data from the primary server. Thus, if/when primary server fails, the secondary server can get elected at that point as the primary to get elected as the leader. (pls note though- it is possible that the follower is "mostly" up to date with primary but might lag behind too . )

Clustered Setup - is typically used when the need is of "High Availability" alongwith "Scalability". In other words, if you need a 2 or more exact copies of data available at any time and also the total dataset size is too big to fit on 1 single primary server (thereby needing Sharding across multiple machines), then a Cluster would be the right way to go.

Foxx is available in both deployment options. 

Do you have any services deployed in Foxx?  What is your projected overall transactions (inserts, updates, lookups) volume? 

Also, for further inputs on your AQL performance question above, @jsteemann might perhaps be able to share more insight.  (fyi - the data dump you shared in one the replies above only contains the vertex collection but no edge collection. were your referring to a different dataset?  If you can clarify more in that context on the Edge inserts/updates point you mentioned above that would be helpful so we understand better)

**Comment by jsteemann** [comment]
@kunalak @epluribusunix : looking the above conversation, I can add the following things:
In a cluster setup, the factors that will influence performance most will be:
* replicationFactor: determines how many copies of each document to keep around. the higher the replicationFactor, the higher the write-performance hit and disk usage requirements. This is obviously unavoidable, but I still want to make you aware of it
* numberOfShards: in general, the more shards a collection has, the more cluster-internal communication there will be when doing full scans or joins. 

As @dothebart mentioned before, a query such as
```
FOR c IN BlueCat
  FOR p in BlueCat
   FILTER p._key == c.parent_id
....
```
without any extra filter conditions will always have to do a full scan in one of the FOR loops, no matter what indexes have been created. If some further filter conditions can be applied that will use an index on the other FOR loop, this may actually help with performance two-fold, by
* using the extra filter to reduce the number of documents to process 
* using an index to only fetch the indexed attribute from the index, in contrast to fetching the full document(s) in a full collection scan. This may reduce the costs a lot if the documents in question are huge. If document sizes aren't too big, this should not have a drastic effect however

Apart from all this, using a streaming cursor may actually help here. Streaming cursors avoid building up the entire query result set in the coordinator's memory before sending the first results to the client application. Instead, when using a streaming cursor the first 1000 results are sent back to the client as soon as they become available. This should reduce the initial response time and also the memory usage on the coordinator.

Not being an expert with arangojs myself, I think the way to make a query use a streaming cursor is like this:
```js
const cursor = await db.query(queryString, bindVars, { options: { stream: true } });
...
```

**Comment by kunalak** [comment]
Hi @epluribusunix  Kerry - how is everything working out for you now after the queries/updates (repserts) re-factoring per our last conversation and with the active-failover deployment mode?  All looking good?

**Comment by epluribusunix** [comment]
@jsteemann Thanks for the tip about streaming cursors - I have implemented them where I can and have found good results, but it does not helped in this specific case.  From what I understand, the memory pressure and processing time comes from the exponential nature of the query, even if no results are being returned.  The problem is the DB engine (on the primary servers) is under pressure and crashing (being killed), not the coordinator or Foxx services.

Again, because I am creating edges for each document, _I really **do** want to iterate through them all._  I don't understand why using an index on the initial/top collection would help with this when the entire collection would be gone through anyway (I cannot exclude, so why filter?).

I also believe that I had the number of shards set too high.  With only two primary DB servers, 8 might have been too much for these relatively low-memory systems to handle.  I would have been better-off with 4, though I did not get to do much testing with this so I can't quantify the impact.

@kunalak The active-failover setup seems to be working quite well.  I re-wrote my DB connection functions so that each connection/call will look for the current leader, and this seems to be resilient enough for my needs.  Ultimately, I would like to configure a proxy server, but the nginx feature I need is enterprise-only, so I  might be forced to research Apache, too.

Although I am not pleased that ETL-type queries seem to be incompatible with clustering, overall, the active-failover setup should work for us, at least for the short-term.  I will revisit clustering in the future, as I agree that monolithic systems are a huge hindrance to growth, but a single server will require far less time/energy to implement and test.  Hopefully, I will have more time to investigate this later this year, but I am far too busy at the moment to spend another week (or two) analyzing cluster performance when it appears the single-server deployment will do the work.

Since my application is definitely query-heavy, I would imagine the worst case having to implement replication partners as query-only slaves - but we'll cross that bridge when we come to it.

**Comment by epluribusunix** [comment]
I think I've finally solved this issue, albeit in a very different way than I expected.

As my usage of the DB increased, I found that queries that modified a large number of documents in one transaction were causing other issues, including a leader/follower switch in the middle of running the query.

My solution was to rewrite the `UPSERT` into a Foxx service that did all the comparison and insert/replace logic in JavaScript.  Now, instead of a single (simple) AQL statement, I look-up both documents (temporary and permanent), compare them using `assert`, and perform a single transaction per insert/replace.

The downside is the extra overhead of many, _**many**_ transactions, but it should scale on a cluster very well, which I look forward to testing again.
